velox/docs/develop/operators.rst

```diff
@@ -232,13 +232,54 @@ followed by the group ID column. The type of group ID column is BIGINT.
      - Description
    * - groupingSets
      - List of grouping key sets. Keys within each set must be unique, but keys can repeat across the sets.
+     - Grouping keys are specified with their output names.
    * - groupingKeyInfos
      - The names and order of the grouping key columns in the output.
    * - aggregationInputs
      - Input columns to duplicate.
    * - groupIdName
      - The name for the group-id column that identifies the grouping set. Zero-based integer corresponding to the position of the grouping set in the 'groupingSets' list.
 
+To illustrate why GroupingSets should use output column names lets examine the following SQL query:
+.. code-block:: sql
+    select COUNT(orderkey), count(distinct orderkey) from orders as t(orderkey);
+
+In this query the user wants to compute both the COUNT of orderkeys and the COUNT of distinct orderkeys. The aggregations
+are related and can be computed in a staged way using multiple grouping sets using the same input column as grouping key.
+Presto generates the following plan for the query:
+
+     - Output[PlanNodeId 11][_col0, _col1] => [count:bigint, count_3:bigint]
```

Comment:
This is not very readable. It would be nice to explain how Presto uses grouping sets to compute this query and provide a sketch of a query plan.

Reply:
> select COUNT(orderkey), count(distinct orderkey) from orders

I'm just trying to understand how Presto plans this query.

It first uses GroupId node to duplicate every record and assign one copy to group 0 and another copy to group 1.

Let's say we have 100 rows with orderkeys 1...100:

```
1
2
...
100
```

After GroupIdNode we get 200 rows

```
k1 k2 gId
1  null 0
2  null 0
...
100 null 0
null 1 1
null 2 1
...
null 100 1 
```

Then, Presto plans partial + final aggregation with grouping key k2, gid and count(k1) aggregate. This produces 101 rows.

```
k2 gid c
null 0 100
1 1 null
2 1 null
...
100 1 null
```

Then, Presto plans another aggregation with no grouping keys and two aggregates: count(k2) and arbitrary(c). This produces 1 row.

```
count(k2) arbitrary(c)
100 100
```

Both count and arbitrary functions ignore null inputs, hence, the results are correct.

I imagine that the same result could be achieved without a GroupId node. First, group by orderkey and count(). Then sum(count) + count.

I wonder why Presto uses GroupId node in this case. Is there another scenario which illustrates that it is necessary? Am I missing anything?

Reply:
For this specific query, the plan was generated when optimize-mixed-distinct-aggregations = true. While this might seem specific, it was a better example to illustrate than some other SQL.

In the tests there was also this example which used GroupId as directed.
SELECT lna, lnb, SUM(quantity) FROM (SELECT linenumber lna, linenumber lnb, CAST(quantity AS BIGINT) quantity FROM lineitem) GROUP BY GROUPING SETS ((lna, lnb), (lna), (lnb), ()) 
Don't know if this example is better.
wdyt ?

Reply:
Yes, the GROUPING SETS example seemed contrived.

I see what kind of explanation you are suggesting. Will do an update.